tensorflow/core/kernels/gather_nd_op_cpu_impl.h

```diff
@@ -113,10 +113,25 @@ struct GatherNdSlice<CPUDevice, T, Index, IXDIM> {
 #endif
     generator::GatherNdSliceGenerator<T, Index, IXDIM> gather_nd_generator(
         slice_size, Tindices, Tparams, Tout, &error_loc);
+
+#ifdef INTEL_MKL
+    // Eigen implementation below is not highly performant. gather_nd_generator
+    // does not seem to be called in parallel, leading to very poor performance.
+    // Additionally, since it uses scalar (Tscratch) to invoke 'generate', it
+    // needs to go through redundant operations like 'reshape', 'broadcast' and
+    // 'sum'. OpenMP loop below essentially does same thing as Eigen code, but
+    // is considerably more efficient.
+    #pragma omp parallel for
+    for (Eigen::DenseIndex i = 0; i < batch_size; i++) {
+      const Eigen::array<Eigen::DenseIndex, 1> loc = i;
+      gather_nd_generator(loc);
+    }
+#else
```

Comment:
thanks!  if you have time, can you also make a non-intel mkl modification below to use Shard(), like [this example](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/reverse_op.cc#L81)?

Reply:
@ebrevdo Sorry, but I don't think I have time for that change. Context switch to make this small change is going to be costly for me. Sorry again.